### PR TITLE
Solution: Stateful clones

### PR DESCRIPTION
**Pull request by stunnerboyone**

### CODE REVIEW

File: src/transformStateWithClones.js [hunk 1]
```diff
@@ -8,6 +8,36 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const clonedState = [];
+  const currentStateData = { ...state };
+  const addDataToArray = () => {
+    clonedState.push({ ...currentStateData });
+  };
```
Comment (Andriy-Shymkiv) [inline src/transformStateWithClones.js:15]: jesus, is that chatgpt helped you achieve this? 😄

File: src/transformStateWithClones.js [hunk 2]
```diff
@@ -8,6 +8,36 @@
...
+
+  for (const action in actions) {
```
Comment (Andriy-Shymkiv) [inline src/transformStateWithClones.js]: `actions` is array. so what loop we should use?

File: src/transformStateWithClones.js [hunk 3]
```diff
@@ -8,6 +8,36 @@
...
+    switch (actions[action].type) {
+      case 'addProperties':
+        Object.assign(currentStateData, { ...actions[action].extraData });
+        addDataToArray();
+        break;
+      case 'removeProperties':
+        for (const keyToRemove in actions[action].keysToRemove) {
+          delete currentStateData[actions[action].keysToRemove[keyToRemove]];
+        }
```
Comment (Andriy-Shymkiv) [inline src/transformStateWithClones.js]: this loop is to complicated, and you can make it simpler

**Review by Andriy-Shymkiv — CHANGES_REQUESTED** [review]
left comments

File: src/transformStateWithClones.js (hunk 1 above)
Comment (stunnerboyone) [reply src/transformStateWithClones.js:15]: No, just watching too many bad youtube channels I think)
I can't use the arrow function or what's wrong here?

File: src/transformStateWithClones.js (hunk 3 above)
Comment (stunnerboyone) [reply src/transformStateWithClones.js]: already Fixed in next commit

File: src/transformStateWithClones.js [hunk 4]
```diff
@@ -8,6 +8,37 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const clonedState = [];
+  const currentStateData = { ...state };
+
+  function addDataToArray() {
+    clonedState.push({ ...currentStateData });
+  };
```
Comment (xiimiik) [inline src/transformStateWithClones.js:16]: we don't need to do this, we only need to do it once in each iteration, immediately AFTER switch

File: src/transformStateWithClones.js [hunk 5]
```diff
@@ -8,6 +8,37 @@
...
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(currentStateData, { ...action.extraData });
```
Comment (xiimiik) [inline src/transformStateWithClones.js:21]: no need to use copy here

```suggestion
        Object.assign(currentStateData, action.extraData );
```

**Review by xiimiik — CHANGES_REQUESTED** [review]
Nice work, a little bit left

File: src/transformStateWithClones.js (hunk 4 above)
Comment (stunnerboyone) [reply src/transformStateWithClones.js:16]: Thanks, got it!

**Review by xiimiik — APPROVED** [review]
GJ 🔥

**Pull request closed without merging**